test/tla/ChangRobertsTyped_Test.tla

```diff
@@ -0,0 +1,125 @@
+--------------------- MODULE ChangRobertsTyped_Test -------------------------
+(*
+ * A test setup for ChangRobertsTyped.
+ *)
+EXTENDS Integers, Apalache
+
+\* a copy of constants from ChangRobertsTyped
+CONSTANTS
+    \* @type: Int;
+    N,
+    \* @type: Int -> Int;
+    Id
+
+\* a copy of state variables from ChangRobertsTyped
+VARIABLES
+    \* @type: Int -> Set(Int);
+    msgs,
+    \* @type: Int -> Str;
+    pc,
+    \* @type: Int -> Bool;
+    initiator,
+    \* @type: Int -> Str;
+    state
+
+INSTANCE ChangRobertsTyped
+
+\* We bound N in the test
+MAX_N == 6
+
+\* we override Node, as N is not known in advance
+OVERRIDE_Node == { i \in 1..MAX_N: i <= N }
+
+\* initialize constants
+ConstInit ==
+    /\ N \in 2..MAX_N
+    /\ Id \in [ 1..MAX_N -> Int ]
+
+\* The below constraints are copied from ASSUME.
+\* They are not enforced automatically, see issue #69.
+Assumptions ==    
+    /\ Node = DOMAIN Id
+    /\ \A n \in Node: Id[n] >= 0
+    /\ \A m,n \in Node : m # n => Id[m] # Id[n]  \* IDs are unique
+
+InitAndAssumptions ==
+    Init /\ Assumptions
+
+(********************** STATELESS OPERATOR TESTS ******************************)
+
+\* Note that succ(n) is not referring to state variables,
+\* so we can test it in isolation.
+\*
+\* @require("ConstInit")
+\* @testStateless
+Test_succ ==
+    \* This is like a property-based test.
+    \* Note that it is exhaustive (for the range of N).
+    \A n \in Node:
+        succ(n) \in Node
+
+\* We could also use a generator, though it is not very useful in case of integers.
+\*
+\* @testStateless
+Test_succ_gen ==
+    \* A property-based test with generators.
+    \* That looks like the canonical form of PBT.
+    \A n \in { Gen(1) }:
+       (n \in Node) => succ(n) \in Node
+
+(*************************** ACTION TESTS *************************************)
+
+\* Preparing the inputs for the test. Note that this is a step of its own.
+\* This is similar to an initialization predicate.
+Prepare_n0 ==
+    \* the following constraint should be added automatically in the future
+    /\ Assumptions
+    \* let the solver pick some data structures within the bounds
+    \* up to 15 messages
+    /\ msgs = Gen(3 * MAX_N)
+    /\ pc = Gen(MAX_N)
+    /\ initiator = Gen(MAX_N)
+    /\ state = Gen(MAX_N)
+    \* restrict the contents with TypeOK,
+    \* so we don't generate useless data structures
+    /\ TypeOK
+
+\* Assertion that we expect to hold true after firing Action_n0.
+Assert_n0 ==
+    \E n, m \in Node:
+        msgs'[n] = msgs[n] \union {m}
+
+\* Execute the action under test.
+\* Note that we decouple Assert_n0 from TestAction_n0.
+\* The reason is that we always assume that TestAction_n0 always holds,
+\* whereas we may want to see Assert_n0 violated.
+\*
+\* @require("ConstInit")
+\* @require("Prepare_n0")
+\* @ensure("Assert_n0")
+\* @testAction
+TestAction_n0 ==
+    \E self \in Node:
+        n0(self)
```

Comment:
> If it is essential, we should maybe work out a coherent syntax for the annotations. E.g., the annotations here don't really fit with the type annotations: here we're missing a trailing ; and we use a quoted argument, rather than a colon an unquoted term.

We have two kinds of syntax for single-argument string annotations: https://apalache.informal.systems/docs/adr/004adr-annotations.html

Reply:
> I ask because, imo, it's generally not great when a parallel system of annotations starts to emerge, and preferable to express things within the language as much as possible. Of course, if it's really essential, then worth pursuing.

I know what you mean. Several times we tried to express a DSL in TLA+ itself, and it happened to be a disaster. I think there are two reasons for that: (1) TLA+ is not an imperative language, (2) formulas have a very flexible structure, so it is tempting to combine things in unexpected ways.

Another reason for having `@require` and `@ensure` in annotations is that you can see them like that in contracts for verification tools, e.g., see Dafny, Ivy, VeriFast. So I am just copying what works well :-)

Reply:
I see! Thanks for explaining. This seems well motivated to me.

I'd be in favor of adding a section to the RFC to record the motivation for putting this in annotations instead of as a DSL.